File: server/qe/sample-frameworks/java-selenium-pom/src/main/java/com/nat/pages/LoginPage.java
package com.nat.pages;

import com.nat.base.BasePage;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;

/**
 * Page Object for the Login page.
 * Demonstrates the POM pattern: locators as fields, actions as methods.
 *
 * <p>Usage:
 * <pre>
 * LoginPage loginPage = new LoginPage(driver);
 * loginPage.login("user@example.com", "password123");
 * </pre>
 */
public class LoginPage extends BasePage {

    // ── Locators ──────────────────────────────────────────────────────────────

    @FindBy(id = "username")
    private WebElement usernameField;

    @FindBy(id = "password")
    private WebElement passwordField;

    @FindBy(css = "button[type='submit']")
    private WebElement loginButton;

    @FindBy(css = ".error-message")
    private WebElement errorMessage;

    @FindBy(css = ".flash-message, .alert-success")
    private WebElement successMessage;

    // ── Constructor ───────────────────────────────────────────────────────────

    /**
     * Constructs LoginPage and initialises @FindBy elements via PageFactory.
     *
     * @param driver the WebDriver for this page
     */
    public LoginPage(WebDriver driver) {
        super(driver);
        PageFactory.initElements(driver, this);
    }

    // ── Actions ───────────────────────────────────────────────────────────────

    /**
     * Enters a value into the username input field.
     *
     * @param username the username to enter
     */
    public void enterUsername(String username) {
        fillInput(usernameField, username);
    }

    /**
     * Enters a value into the password input field.
     *
     * @param password the password to enter
     */
    public void enterPassword(String password) {
        fillInput(passwordField, password);
    }

    /**
     * Clicks the login submit button.
     */
    public void clickLogin() {
        clickElement(loginButton);
    }

    /**
     * Returns true if an error message element is visible.
     *
     * @return true if error is displayed
     */
    public boolean isErrorMessageVisible() {
        return isElementVisible(errorMessage);
    }

    /**
     * Returns the visible text of the error message element.
     *
     * @return the error message string
     */
    public String getErrorMessage() {
        return getText(errorMessage);
    }

    /**
     * Returns the error message WebElement directly.
     * Used by AssertUtils for element-level assertions.
     *
     * @return the error message WebElement
     */
    public WebElement getErrorMessageElement() {
        return errorMessage;
    }

    /**
     * Convenience method: enters username and password, then clicks login.
     *
     * @param username the username
     * @param password the password
     */
    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
    }
}

File: server/qe/sample-frameworks/typescript-playwright-pom/src/pages/LoginPage.ts
import { type Page, type Locator } from '@playwright/test';
import { BasePage } from '../base/BasePage';

export interface LoginData {
  username: string;
  password: string;
}

/**
 * Page Object for the Login page.
 * All locators are defined as readonly class properties.
 *
 * @example
 * ```typescript
 * const loginPage = new LoginPage(page);
 * await loginPage.navigate('/login');
 * await loginPage.loginWith({ username: 'user@example.com', password: 'pass' });
 * ```
 */
export class LoginPage extends BasePage {
  // ── Locators ────────────────────────────────────────────────────────────────
  readonly usernameField: Locator;
  readonly passwordField: Locator;
  readonly loginButton: Locator;
  readonly errorMessage: Locator;
  readonly successMessage: Locator;
  readonly rememberMeCheckbox: Locator;
  readonly forgotPasswordLink: Locator;

  constructor(page: Page) {
    super(page);
    this.usernameField      = page.locator('#username, input[name="username"], input[type="email"]').first();
    this.passwordField      = page.locator('#password, input[name="password"], input[type="password"]').first();
    this.loginButton        = page.locator('button[type="submit"], input[type="submit"], button:has-text("Login"), button:has-text("Sign In")').first();
    this.errorMessage       = page.locator('.error-message, .alert-danger, [data-testid="error-message"]').first();
    this.successMessage     = page.locator('.flash-message, .alert-success, [data-testid="success-message"]').first();
    this.rememberMeCheckbox = page.locator('input[type="checkbox"][name*="remember"]').first();
    this.forgotPasswordLink = page.locator('a:has-text("Forgot"), a:has-text("Reset")').first();
  }

  // ── Actions ─────────────────────────────────────────────────────────────────

  /**
   * Fills username and password fields, then clicks the submit button.
   */
  async loginWith(data: LoginData): Promise<void> {
    await this.enterUsername(data.username);
    await this.enterPassword(data.password);
    await this.submit();
  }

  async enterUsername(value: string): Promise<void> {
    await this.fillInput(this.usernameField, value);
  }

  async enterPassword(value: string): Promise<void> {
    await this.fillInput(this.passwordField, value);
  }

  async submit(): Promise<void> {
    await this.clickElement(this.loginButton);
    await this.waitForNavigation().catch(() => {
      // navigation may not occur on failed login — that is expected
    });
  }

  async getErrorMessage(): Promise<string> {
    return this.getText(this.errorMessage);
  }

  async isErrorVisible(): Promise<boolean> {
    return this.isVisible(this.errorMessage);
  }
}
